### PR TITLE
Http Instrumentation: Fix gRPC causing issues retrieving request & response objects

### PR DESCRIPTION
**Pull request by CodeBlanch**
Fixes https://github.com/open-telemetry/opentelemetry-dotnet/issues/2652

## Changes

* If `PropertyFetcher` is given a different payload type than what was used during construction, it will now build an inner instance and defer to that. We could have also looked up the correct `PropertyFetcher` from some kind of dictionary (by payload type), but I went with this design to keep the common path (where only a single payload type is used) fast/lock-free.

## TODOs

* [X] `CHANGELOG.md` updated for non-trivial changes
* [X] Unit test

### CODE REVIEW

File: test/OpenTelemetry.Tests/Instrumentation/PropertyFetcherTest.cs [hunk 1]
```diff
@@ -50,5 +50,35 @@ public void FetchNullProperty()
...
+        [Fact]
+        public void FetchPropertyMultiplePayloadTypes()
+        {
+            var fetch = new PropertyFetcher<string>("Property");
+
+            Assert.True(fetch.TryFetch(new PayloadTypeA(), out string propertyValue));
+            Assert.Equal("A", propertyValue);
+
+            Assert.True(fetch.TryFetch(new PayloadTypeB(), out propertyValue));
+            Assert.Equal("B", propertyValue);
```
Comment (alanwest) [inline test/OpenTelemetry.Tests/Instrumentation/PropertyFetcherTest.cs:63]: Interesting, so if most of the fetches are against `PayloadTypeB` but the first invocation for a process is against `PayloadTypeA` there'll always be a fetch first assuming `A` and then another for `B`. Probably not a huge deal overhead-wise, but just explicitly calling it out.

**Review by alanwest — APPROVED** [review]
LGTM

File: test/OpenTelemetry.Tests/Instrumentation/PropertyFetcherTest.cs (hunk 1 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/Instrumentation/PropertyFetcherTest.cs:63]: Ya if you look at the issue it looks like some kind of Azure Functions thing where it always does a gRPC call first. So for native HttpClient invocations in that environment, it is going to pay the penalty of a failed `is` check and then an extra virtual dispatch on each call. Very minor hit, but a hit nonetheless. Not much we can do about it. If we instead did a dictionary lookup, we would have to pay for a hash + lookup for _all_ invocations even in cases where only one payload is ever used. So this seemed like the lesser evil. Kind of lame this gRPC path uses the same diagnostic source but with different payload types.

File: src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+* Fixed an issue with `Filter` and `Enrich` callbacks not firing under certain
+  conditions when gRPC is used
+  ([#2698](https://github.com/open-telemetry/opentelemetry-dotnet/issues/2698))
```
Comment (mic-max) [inline src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:7]: Should be:
([#2698](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2698))

Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.Http/CHANGELOG.md:7]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/2750

**Pull request merged**